drivers/regex/mlx5/mlx5_regex_queue.c

```diff
@@ -0,0 +1,263 @@
+	seg->lkey = rte_cpu_to_be_32(lkey);
+	seg->addr = rte_cpu_to_be_64(address);
+}
+
+/**
+ * Send a single request
+ *
+ * @param dev
+ *   Pointer to REGEX device.
+ * @param q
+ *   The queue to send the request.
+ * @param ops
+ *   Pointer to the request.
+ *
+ * @return
+ *   0: success else error code.
+ */
+static int
+tx_burst_msegs_op(struct rte_regex_dev *dev, struct mlx5_regex_qp *q,
+		  struct rte_regex_ops *ops)
+{
+	struct mlx5_wqe_metadata_seg *restrict mseg;
+	struct mlx5_regex_wqe_ctrl_seg regex_cseg;
+	struct mlx5_wqe_data_seg *restrict dseg;
+	struct mlx5_wqe_ctrl_seg *restrict cseg;
+	struct mlx5_wqe_data_seg tmp_dseg;
+	int wqe_idx = q->sq.pi % q->sq.num_entries;
+	struct mlx5_regex_job *job;
+	size_t wqe_offset;
+	uint32_t lkey;
+	int nsegs;
+
+	if (q->sq.pi - q->sq.ci >= q->sq.num_entries)
+		return EBUSY;
+	nsegs = ops->num_of_bufs;
+	/*
+	 * REGEX dev currently support single segment req buf.
+	 * TODO: UMR to support multiple segments.
+	 */
+	assert(nsegs == 1);
+	/*
+	 * WQE:
+	 * - 1 Control Segment
+	 * - 1 Metadata Segment
+	 * - 1 Data Segment for request
+	 * - 1 Data Segment for response
+	 */
+	nsegs += 3;
+	wqe_offset = wqe_idx * MLX5_SEND_WQE_BB ;
+	job = q->regex_jobs[wqe_idx];
+	job->in_progress = true;
+	job->user_id = ops->user_id;
+	cseg = (struct mlx5_wqe_ctrl_seg *)
+		((uint8_t *)(uintptr_t)q->sq.umem_buf + wqe_offset);
+	/* TODO: ask Yuval about the last param ctrl_field */
+	set_ctrl_seg((void *)&regex_cseg, 0, ops, /* ctrl_field */ 0);
+#ifdef MLX5_REGEX_REAL_HW
+	mlx5dv_set_ctrl_seg(cseg, q->sq.pi & MLX5_QP_CI_MASK, MLX5_OPCODE_MMO,
+			    MLX5_OPC_MOD_MMO_REGEX, q->sq.sqn,
+			    MLX5_WQE_CTRL_CQ_UPDATE, nsegs, 0,
+			    regex_cseg.le_subset_id_0_subset_id_1);
+#else
+	mlx5dv_set_ctrl_seg(cseg, q->sq.pi & MLX5_QP_CI_MASK, MLX5_OPCODE_NOP,
+			    0, q->sq.sqn, MLX5_WQE_CTRL_CQ_UPDATE, 1, 0, 0);
+#endif
+	/* Set wqe metadata segment */
+	mseg = (struct mlx5_wqe_metadata_seg *)((uint8_t*)cseg + sizeof(*cseg));
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					(uintptr_t)BUF_ADDR(&job->m_iov));
+	mlx5dv_set_data_seg(&tmp_dseg, BUF_SIZE(&job->m_iov),
+			    lkey, (uintptr_t)BUF_ADDR(&job->m_iov));
+	mlx5dv_set_metadata_seg(mseg, regex_cseg.ctrl_subset_id_2_subset_id_3,
+				lkey, tmp_dseg.addr);
+	/* Data segment for nsegs buffer */
+	dseg = (struct mlx5_wqe_data_seg *)((uint8_t*)mseg + sizeof(*mseg));
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					(uintptr_t)BUF_ADDR((*ops->bufs)[0]));
+	mlx5dv_set_data_seg(dseg, BUF_SIZE((*ops->bufs)[0]),
+			    lkey, (uintptr_t)BUF_ADDR((*ops->bufs)[0]));
+	/* Data segment for response buffer */
+	dseg += sizeof(*dseg);
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					(uintptr_t)BUF_ADDR(&job->resp_iov));
+	mlx5dv_set_data_seg(dseg, BUF_SIZE(&job->resp_iov),
+			    lkey, (uintptr_t)BUF_ADDR(&job->resp_iov));
+	/* Ring doorbell */
+	rte_io_wmb();
+	q->sq.db_rec[MLX5_SND_DBR] = rte_cpu_to_be_32(q->sq.pi & MLX5_QP_CI_MASK);
```

Comment:
Hi vu
this a bug
please see PRM
need to be the next empty wqbb and not wqe index